File: src/main/java/de/kleiner3/lasertag/lasertaggame/settings/SettingDataType.java
package de.kleiner3.lasertag.lasertaggame.settings;

/**
 * @author Étienne Muser
 */
public class SettingDataType {
    public static final SettingDataType LONG = new SettingDataType(Long.class);
    public static final SettingDataType BOOL = new SettingDataType(Boolean.class);

    private final Class<?> valueType;

    private final boolean isEnum;

    /**
     * This constructor is only defined to hide the default constructor. DO NOT USE!
     */
    @Deprecated
    private SettingDataType() {
        this(null, false);
    }

    private SettingDataType(Class<?> valueType) {
        this(valueType, false);
    }

    private SettingDataType(Class<?> valueType, boolean isEnum) {
        this.valueType = valueType;
        this.isEnum = isEnum;
    }

    public Class<?> getValueType() {
        return this.valueType;
    }

    public boolean isEnum() {
        return this.isEnum;
    }

    public static SettingDataType ofEnum(Class<? extends Enum<?>> enumType) {
        return new SettingDataType(enumType, true);
    }
}

File: pewpewcubes-entities/src/main/kotlin/de/pewpewproject/pewpewcubes/entities/settings/SettingDataType.kt
package de.pewpewproject.pewpewcubes.entities.settings

/**
 * The possible data types for settings
 * @author Étienne Muser
 */
class SettingDataType private constructor(val valueType: Class<*>?, val isEnum: Boolean = false) {
    @Deprecated("This constructor is only defined to hide the default constructor. DO NOT USE!")
    private constructor() : this(null, false)

    companion object {
        val LONG: SettingDataType = SettingDataType(Long::class.java)
        val BOOL: SettingDataType = SettingDataType(Boolean::class.java)

        fun ofEnum(enumType: Class<out Enum<*>?>?): SettingDataType {
            return SettingDataType(enumType, true)
        }
    }
}